### PR TITLE
implement `Metadata::retain`

### LINKED ISSUE #10683: Add `Metadata::retain` to filter entries in place
**Is your feature request related to a problem or challenge?**
- Found while testing upgrade of arrow in DataFusion: https://github.com/apache/datafusion/pull/24366
- Related to https://github.com/apache/arrow-rs/pull/10075

While updating DataFusion to arrow 60 I  found  the new `Metadata` struct  introduced in https://github.com/apache/arrow-rs/pull/10075 has no way to remove entries based on a predicate. Code that previously called `HashMap::retain` on field/schema metadata now has to convert to a `BTreeMap`, filter, and convert back.

For example, DataFusion computes the intersection of metadata across UNION inputs, which now looks like:

```rust
let mut intersected: Option<BTreeMap<String, String>> = None;

for metadata in metadatas {
    match &mut intersected {
        None => {
            // deep copy into a BTreeMap so we can filter it below
            intersected =
                Some(metadata.iter().map(|(k, v)| (k.clone(), v.clone())).collect());
        }
        Some(current) => {
            current.retain(|k, v| metadata.get(k) == Some(&*v));
        }
    }
}

intersected.map(Metadata::from).unwrap_or_default()
```

**Describe the solution you'd like**

A `Metadata::retain(|k, v| ...)` method, copy-on-write like the existing `insert`/`remove`. The code above would then stay in terms of `Metadata` and avoid the deep copy and round-trip:

```rust
let mut intersected: Option<Metadata> = None;

for metadata in metadatas {
    match &mut intersected {
        None => intersected = Some(metadata.clone()), // cheap clone
        Some(current) => current.retain(|k, v| metadata.get(k) == Some(&*v)),
    }
}

intersected.unwrap_or_default()
```

**Describe alternatives you've considered**

Converting through `BTreeMap` as shown above; it works but adds a copy and obscures the intent.

**Additional context**

### PR DESCRIPTION
**Pull request by Rich-T-kid**
# Which issue does this PR close?

- Closes #10683.

# Rationale for this change

see #10683

# What changes are included in this PR?
implements `retain` on metadata. The underlying BTreeMap is allocation free unless the `Arc` pointer is shared

# Are these changes tested?
yes

# Are there any user-facing changes?
yes, new `retain` method on the metadata struct

### CODE REVIEW

File: arrow-schema/src/metadata.rs [hunk 1]
```diff
@@ -146,6 +146,19 @@ impl Metadata {
...
+    ///
+    /// Entries for which `f` returns `false` are removed. Retained entries
+    /// remain in their original (sorted) order.
+    ///
+    /// Clones the underlying map if (and only if) it is shared.
+    pub fn retain<F>(&mut self, eq: F)
+    where
+        F: FnMut(&String, &mut String) -> bool,
+    {
+        Arc::make_mut(self.0.get_or_insert_default()).retain(eq);
```
Comment (alamb) [inline arrow-schema/src/metadata.rs]: Does calling get_or_insert_default mean this method would turn a metadata that had a None inner map into aSome(empty map)?

I think it would make more sense to do nothing if `self.0` is None

File: arrow-schema/src/metadata.rs [hunk 2]
```diff
@@ -445,6 +458,35 @@ mod tests {
...
+        let mut metadata =
+            Metadata::from([("a", "1"), ("b", "2"), ("c", "3"), ("d", "4"), ("e", "5")]);
+
+        metadata.retain(|k, _| -> bool { k >= &String::from("c") });
+
+        let result_map = Metadata::from([("c", "3"), ("d", "4"), ("e", "5")]);
+        assert_eq!(metadata, result_map)
+    }
+
+    #[test]
```
Comment (alamb) [inline arrow-schema/src/metadata.rs:492]: Can you please add a test that shows what happens when retain is called on a empty Metadata (`None`)? It should I think still be `None`

**Review by alamb — COMMENTED** [review]
Thank you @Rich-T-kid  -- I think this is very close -- I just have one question

File: arrow-schema/src/metadata.rs (hunk 1 above)
Comment (Rich-T-kid) [reply arrow-schema/src/metadata.rs]: ah good catch!

File: arrow-schema/src/metadata.rs (hunk 2 above)
Comment (Rich-T-kid) [reply arrow-schema/src/metadata.rs:492]: also added a test for when `retain()` clears all the maps contents. the metadata becomes `None`

**Review by alamb — APPROVED** [review]
Thank you @Rich-T-kid
